### ISSUE TITLE
Make selected tests clean up after themselves

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1883]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#1883|
|Type|task|
|Created|2006-06-30 14:52:56Z|

Trial's clean up code spins the reactor a couple of times before cleaning up pending calls.  This won't work if the tests are run in a reactor.  Removing the iterate calls make a whole bunch of tests spit out cleanup errors. 

<details><summary>Searchable metadata</summary>

```
trac-id__1883 1883
type__task task
reporter__jml jml
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1151679176000000 1151679176000000
changetime__1158895969000000 1158895969000000
version__None None
owner__ 
cc__exarkun cc__jerub
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Roughly 120+ Unclean Reactor warnings when the lines are removed.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

OK. So maybe this can't be done by cleaning up tests, maybe the problem is best fixed in the implementation of do_cleanPending.

It can't call reactor.iterate(0), because it needs to be run in a callback.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

I would suggest making it do reactor.callFromThread(finish) instead. That is pretty much an equivalent of reactor.iterate(0).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```reactor.callLater(0, ...)``` twice should exactly duplicate the current behavior. I think using ```callLater``` is a little less obscure than using ```callFromThread```.  ```callFromThread``` also isn't available on platforms without threading support.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

I agree that it should, but it doesn't, due to the callLater ordering bug ([#6718](https://github.com/twisted/twisted/issues/6718)) that I said I would fix but have not merged. =/

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Does any code actually depend on that difference?  It's pretty subtle, and the point of the iterate() calls are just to get rid of anything that's _almost_ done.

Ultimately, things which are still active at this point in a test should probably just be considered failures, so we just need a solution that's good enough to get us far enough to start fixing those things.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Jp, what do you mean by "good enough to get us far enough to start fixing those things". After all, removing the iterate calls makes it fairly clear which tests are failing.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Difference in ordering? Yes, simplified example:
```
# Test case code:
  reactor.callLater(0, lambda: None)
# Trial code:
  reactor.callLater(0, finishCleanUp)
```

Because of the bug, finishCleanUp **sometimes** will get called before frobnitz (depending on system clock resolution and timing). Therefore, if the goal is to just let things almost done finish, that is not a suitably reliable way to do so.

reactor.callLater(0.000001, finishCleanUp), would work, though.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml set owner to @jerub<br>@jml set status to `new`|
|-|-|

Ugh. This branch has been around for ages and is not going anywhere anytime soon.

I'm putting it up for review -- it fixes a couple of dodgy tests, those fixes should get merged.

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|moonfallen commented|
|-|-|

Review:

- commented out the reactor.iterate(0) calls, and unclean reactor appears.  Apply the patch for this bug, and unclean reactor goes away.  Ergo, this patch fixes what it says it fixes.

- Did not introduce any test regressions.

review = ok.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml<br>@jml set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_17"></a>|@jml set status to `closed`|
|-|-|

(In [18220]) Make tests clean up after themselves
 
 * Fixes [#1883](https://github.com/twisted/twisted/issues/1883)
 * Author: jml
 * Reviewer: MFen

Although not apparent to the naked eye, these tests previous did not
correctly clean up after themselves. Instead, they relied on a couple 
of spins of the reactor to do their cleaning for them. This has been
corrected.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
